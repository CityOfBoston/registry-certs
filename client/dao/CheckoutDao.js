// @flow
import { action, runInAction } from 'mobx';

import type { LoopbackGraphql } from '../lib/loopback-graphql';
import type Cart from '../store/Cart';
import type Order from '../models/Order';

import submitDeathCertificateOrder from '../queries/submit-death-certificate-order';

export default class CheckoutDao {
  loopbackGraphql: LoopbackGraphql;
  stripe: ?StripeInstance;

  constructor(loopbackGraphql: LoopbackGraphql, stripe: ?StripeInstance) {
    this.loopbackGraphql = loopbackGraphql;
    this.stripe = stripe;
  }

  // Given a StripeElement object, tokenizes the card info with Stripe and, if
  // successful, populates the cardToken and cardLast4 values in Orderâ€™s info
  // and returns true. Returns false and populates processingError if it fails
  // for network or validation reasons.
  @action
  async tokenizeCard(
    order: Order,
    cardElement: ?StripeElement
  ): Promise<boolean> {
    const { stripe } = this;

    if (!stripe || !cardElement) {
      throw new Error(
        'tokenizeCard called without Stripe and/or StripeElement'
      );
    }

    try {
      order.processing = true;
      order.processingError = null;

      const tokenResult = await stripe.createToken(cardElement, {
        name: order.info.cardholderName,
        address_line1: order.billingAddress1,
        address_line2: order.billingAddress2,
        address_city: order.billingCity,
        address_state: order.billingState,
        address_zip: order.billingZip,
        address_country: 'us',
      });

      if (tokenResult.error) {
        runInAction(
          'CheckoutDao > tokenizeCard > createToken error result',
          () => {
            order.processingError = tokenResult.error.message;
          }
        );

        return false;
      } else {
        const { token } = tokenResult;
        const { card } = token;

        runInAction('CheckoutDao > tokenizeCard > createToken success', () => {
          order.cardToken = token.id;
          order.cardFunding = card.funding;
          order.info.cardLast4 = card.last4;
        });

        return true;
      }
    } catch (err) {
      runInAction('CheckoutDao > tokenizeCard > catch block', () => {
        order.processingError =
          err.message || `Unexpected error submitting order: ${err}`;
      });

      if (window._opbeat && !err._sentToOpbeat) {
        window._opbeat('captureException', err);
      }

      return false;
    } finally {
      runInAction('CheckoutDao > tokenizeCard > finally', () => {
        order.processing = false;
      });
    }
  }

  @action
  async submit(cart: Cart, order: Order): Promise<?string> {
    try {
      order.processing = true;
      order.processingError = null;

      const orderId = await submitDeathCertificateOrder(
        this.loopbackGraphql,
        cart,
        order
      );

      cart.clear();

      return orderId;
    } catch (err) {
      runInAction('CheckoutDao > submit > catch block', () => {
        order.processingError =
          err.message || `Unexpected error submitting order: ${err}`;
      });

      if (window._opbeat && !err._sentToOpbeat) {
        window._opbeat('captureException', err);
      }

      return null;
    } finally {
      runInAction('CheckoutDao > submit > finally', () => {
        order.processing = false;
      });
    }
  }
}
