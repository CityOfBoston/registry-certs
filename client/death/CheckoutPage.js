// @flow

import React from 'react';
import Head from 'next/head';
import { observer } from 'mobx-react';
import Link from 'next/link';

import { getDependencies } from '../app';

import { CERTIFICATE_COST, PROCESSING_FEE } from '../store/Cart';
import type Cart from '../store/Cart';

import AppLayout from '../AppLayout';

import Nav from '../common/Nav';
import CheckoutItem from './checkout/CheckoutItem';

type Props = {
  cart: Cart,
};

@observer
export class CheckoutPageContent extends React.Component<Props> {
  render() {
    const { cart } = this.props;

    return (
      <AppLayout>
        {() => (
          <div>
            <Head>
              <title>Boston.gov — Death Certificate Checkout</title>
            </Head>

            <Nav cart={cart} link="lookup" />

            <div className="p-a300">
              <div className="sh sh--b0 m-v300">
                <h1 className="sh-title">Checkout</h1>
              </div>
            </div>

            <div>
              {cart.items.map(item => (
                <CheckoutItem key={item.id} item={item} cart={cart} />
              ))}

              <div className="p-a300 g">
                {this.renderCost()}

                <Link href="/death/payment">
                  <a className="btn g--3 m-v500">Pay and Finish</a>
                </Link>

                <style jsx>{`
                  a.btn {
                    align-self: center;
                    text-align: center;
                  }
                `}</style>
              </div>
            </div>
          </div>
        )}
      </AppLayout>
    );
  }

  renderCost() {
    const { cart } = this.props;

    return (
      <div className="m-v500 g--9">
        <div className="t--info">
          {cart.size} {cart.size === 1 ? 'certificate' : 'certificates'} × ${CERTIFICATE_COST}{' '}
          + {(PROCESSING_FEE * 100).toFixed(2)}% credit card fee
        </div>
        <div className="sh sh--b0">
          <span className="sh-title">Subtotal: ${cart.cost.toFixed(2)}</span>
        </div>
      </div>
    );
  }
}

export default function CheckoutPageContentController() {
  const { cart } = getDependencies();
  return <CheckoutPageContent cart={cart} />;
}
