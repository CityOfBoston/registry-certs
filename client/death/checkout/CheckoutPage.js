// @flow
// Wrapper controller for ShippingContent and PaymentContent

import React, { type Element as ReactElement } from 'react';
import Router from 'next/router';

import {
  getDependencies,
  type ClientContext,
  type ClientDependencies,
} from '../../app';

import AppLayout from '../../AppLayout';

import Order from '../../models/Order';

import PaymentContent, {
  type Props as PaymentContentProps,
} from './PaymentContent';

import ShippingContent, {
  type Props as ShippingContentProps,
} from './ShippingContent';

import ConfirmationContent, {
  type Props as ConfirmationContentProps,
} from './ConfirmationContent';

type InitialProps =
  | {
      page: 'shipping',
    }
  | {
      page: 'payment',
    }
  | {
      page: 'confirmation',
      orderId: string,
      contactEmail: string,
    };

type CheckoutContentProps =
  | {
      page: 'shipping',
      props: ShippingContentProps,
    }
  | {
      page: 'payment',
      props: PaymentContentProps,
    }
  | {
      page: 'confirmation',
      props: ConfirmationContentProps,
    };

export function wrapCheckoutPageController(
  getDependencies: (ctx?: ClientContext) => ClientDependencies,
  renderContent: (ClientDependencies, CheckoutContentProps) => ?ReactElement<*>
) {
  return class CheckoutPageController extends React.Component<InitialProps> {
    static getInitialProps({ query, res }: ClientContext): InitialProps {
      let props: InitialProps;
      let redirectToShippingIfServer = false;

      switch (query.page || '') {
        case '':
        case 'shipping':
          props = { page: 'shipping' };
          break;
        case 'payment':
          props = { page: 'payment' };

          // We know that we won't have shipping information if someone visits
          // the payment page directly from the server (since this component's
          // internal state will be blank), so redirect.
          redirectToShippingIfServer = true;
          break;
        case 'confirmation':
          props = {
            page: 'confirmation',
            orderId: query.orderId || '',
            contactEmail: query.contactEmail || '',
          };
          break;
        default:
          props = { page: 'shipping' };
          redirectToShippingIfServer = true;
      }

      if (redirectToShippingIfServer && res) {
        res.writeHead(301, {
          Location: '/death/checkout',
        });
        res.end();
        res.finished = true;
      }

      return props;
    }

    dependencies = getDependencies();

    // This wil persist across different sub-pages since React will preserve the
    // component instance. This will keep the form fields filled out as the user
    // goes forward and back in the interface, and it will get automatically
    // disposed of if the user leaves the flow, which is the behavior that we
    // want.
    order: Order;

    componentWillMount() {
      const { orderProvider } = this.dependencies;
      // This will be populated from localStorage, and changes to it will get
      // written back there.
      this.order = orderProvider.get();
    }

    componentDidMount() {
      this.redirectIfMissingShipping(this.props);
    }

    componentWillReceiveProps(newProps: InitialProps) {
      this.redirectIfMissingShipping(newProps);
    }

    // In the case of reloading from the browser, for example, or clicking
    // "back" from the confirmation page.
    async redirectIfMissingShipping(props: InitialProps) {
      if (props.page === 'payment' && !this.order.shippingIsComplete) {
        await Router.push('/death/checkout?page=shipping');
        window.scrollTo(0, 0);
      }
    }

    advanceToPayment = async () => {
      await Router.push('/death/checkout?page=payment');
      window.scrollTo(0, 0);
    };

    submitOrder = async (cardElement: ?StripeElement) => {
      const { order } = this;
      const { cart, checkoutDao } = this.dependencies;

      const orderId = await checkoutDao.submit(cart, order, cardElement);

      if (orderId) {
        this.order = new Order();

        await Router.push(
          `/death/checkout?page=confirmation&orderId=${encodeURIComponent(
            orderId
          )}&contactEmail=${encodeURIComponent(order.info.contactEmail)}`,
          '/death/checkout?page=confirmation'
        );

        window.scrollTo(0, 0);
      }
    };

    render() {
      const { props, order } = this;
      const { cart, stripe } = this.dependencies;

      let renderProps;
      switch (props.page) {
        case 'shipping':
          renderProps = {
            page: 'shipping',
            props: {
              cart,
              order,
              submit: this.advanceToPayment,
            },
          };
          break;
        case 'payment':
          renderProps = {
            page: 'payment',
            props: {
              stripe,
              cart,
              order,
              submit: this.submitOrder,
            },
          };
          break;
        case 'confirmation':
          renderProps = {
            page: 'confirmation',
            props: {
              orderId: props.orderId,
              contactEmail: props.contactEmail,
            },
          };
          break;
        default:
          throw new Error(`Unknown page: ${(props.page: any)}`);
      }

      return renderContent(this.dependencies, renderProps);
    }
  };
}

export default wrapCheckoutPageController(
  getDependencies,
  ({ cart }, props) => {
    switch (props.page) {
      case 'shipping':
        return (
          <AppLayout navProps={null}>
            {React.createElement(ShippingContent, props.props)}
          </AppLayout>
        );
      case 'payment':
        return (
          <AppLayout navProps={null}>
            {React.createElement(PaymentContent, props.props)}
          </AppLayout>
        );
      case 'confirmation':
        return (
          <AppLayout navProps={{ cart, link: 'lookup' }}>
            {React.createElement(ConfirmationContent, props.props)}
          </AppLayout>
        );
      default:
        throw new Error(`Unknown page: ${props.page}`);
    }
  }
);
