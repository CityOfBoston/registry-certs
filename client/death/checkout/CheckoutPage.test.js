// @flow

import Router from 'next/router';

import CheckoutPage, { wrapCheckoutPageController } from './CheckoutPage';
import CheckoutDao from '../../dao/CheckoutDao';
import OrderProvider from '../../store/OrderProvider';

jest.mock('next/router');
jest.mock('../../dao/CheckoutDao');

beforeEach(() => {
  jest.clearAllMocks();
});

describe('integration', () => {
  // We have to componentWillMount in all of these to initialize the
  // Order object.
  it('renders shipping', () => {
    const page = new CheckoutPage({ page: 'shipping' });
    page.componentWillMount();
    expect(page.render()).toMatchSnapshot();
  });

  it('renders payment', () => {
    const page = new CheckoutPage({ page: 'payment' });
    page.componentWillMount();
    expect(page.render()).toMatchSnapshot();
  });

  it('renders confirmation', () => {
    const page = new CheckoutPage({
      page: 'confirmation',
      orderId: '123-456-7',
      contactEmail: 'ttoe@squirrelzone.net',
    });
    page.componentWillMount();
    expect(page.render()).toMatchSnapshot();
  });
});

describe('controller', () => {
  describe('getInitialProps', () => {
    let CheckoutPageController;
    let dependencies: any;
    let res: ServerResponse;

    beforeEach(() => {
      dependencies = {};
      res = {
        writeHead: jest.fn(),
        end: jest.fn(),
        finished: false,
      };

      CheckoutPageController = wrapCheckoutPageController(
        () => dependencies,
        () => null
      );
    });

    it('treats no page query param as shipping', () => {
      const initialProps = CheckoutPageController.getInitialProps(
        ({ res, query: {} }: any)
      );

      expect(initialProps.page).toEqual('shipping');
    });

    it('process payment on server', () => {
      const initialProps = CheckoutPageController.getInitialProps(
        ({ query: { page: 'payment' } }: any)
      );

      expect(initialProps.page).toEqual('payment');
    });

    it('redirects payment on server back to shipping', () => {
      CheckoutPageController.getInitialProps(
        ({ res, query: { page: 'payment' } }: any)
      );

      expect(res.writeHead).toHaveBeenCalled();
      expect(res.finished).toBe(true);
    });

    it('redirects unknown page query param to shipping on server', () => {
      CheckoutPageController.getInitialProps(
        ({ res, query: { page: 'not-a-real-page' } }: any)
      );

      expect(res.writeHead).toHaveBeenCalled();
      expect(res.finished).toBe(true);
    });

    it('treats unknown page query param as shipping on client', () => {
      const initialProps = CheckoutPageController.getInitialProps(
        ({
          query: { page: 'not-a-real-page' },
        }: any)
      );

      expect(initialProps.page).toEqual('shipping');
    });

    it('passes confirm props along', () => {
      const initialProps = CheckoutPageController.getInitialProps(
        ({
          query: {
            page: 'confirmation',
            orderId: '123-456-7',
            contactEmail: 'ttoe@squirrelzone.net',
          },
        }: any)
      );

      expect(initialProps).toEqual({
        page: 'confirmation',
        orderId: '123-456-7',
        contactEmail: 'ttoe@squirrelzone.net',
      });
    });
  });

  describe('operations', () => {
    let dependencies: any;
    let checkoutDao: CheckoutDao;
    let orderProvider: OrderProvider;
    let controller;

    beforeEach(() => {
      orderProvider = new OrderProvider();
      checkoutDao = new CheckoutDao((null: any), null);

      dependencies = {
        checkoutDao,
        orderProvider,
      };

      const CheckoutPageController = wrapCheckoutPageController(
        () => dependencies,
        () => null
      );

      // page doesn't really matter for this
      controller = new CheckoutPageController({ page: 'shipping' });
      controller.componentWillMount();
    });

    describe('advanceToPayment', () => {
      it('routes to next stage', () => {
        controller.advanceToPayment();
        expect(Router.push).toHaveBeenCalled();
      });
    });

    describe('submitOrder', () => {
      it('exists', () => {
        expect(controller.submitOrder).toBeDefined();
      });

      it('redirects on success', async () => {
        checkoutDao.submit.mockReturnValue(Promise.resolve('test-order-id'));
        await controller.submitOrder(null);
        expect(Router.push).toHaveBeenCalled();
      });

      it('stays on the same page on failure', async () => {
        checkoutDao.submit.mockReturnValue(Promise.resolve(null));
        await controller.submitOrder(null);
        expect(Router.push).not.toHaveBeenCalled();
      });
    });
  });
});
