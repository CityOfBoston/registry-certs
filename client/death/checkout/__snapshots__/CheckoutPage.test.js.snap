// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration renders confirmation 1`] = `
<AppLayout
  navProps={
    Object {
      "cart": Cart {},
      "link": "lookup",
    }
  }
>
  <ConfirmationContent
    contactEmail="ttoe@squirrelzone.net"
    orderId="123-456-7"
  />
</AppLayout>
`;

exports[`integration renders payment 1`] = `
<AppLayout
  navProps={null}
>
  <PaymentContent
    cart={Cart {}}
    order={
      Order {
        "cardElementComplete": false,
        "cardElementError": null,
        "info": Object {
          "billingAddress1": "",
          "billingAddress2": "",
          "billingAddressSameAsShippingAddress": true,
          "billingCity": "",
          "billingState": "",
          "billingZip": "",
          "cardholderName": "",
          "contactEmail": "",
          "contactName": "",
          "contactPhone": "",
          "shippingAddress1": "",
          "shippingAddress2": "",
          "shippingCity": "",
          "shippingCompanyName": "",
          "shippingName": "",
          "shippingState": "",
          "shippingZip": "",
          "storeBilling": false,
          "storeContactAndShipping": false,
        },
        "submissionError": null,
        "submitting": false,
        "updateStorageDisposer": null,
      }
    }
    stripe={null}
    submit={[Function]}
  />
</AppLayout>
`;

exports[`integration renders shipping 1`] = `
<AppLayout
  navProps={null}
>
  <ShippingContent
    cart={Cart {}}
    order={
      Order {
        "cardElementComplete": false,
        "cardElementError": null,
        "info": Object {
          "billingAddress1": "",
          "billingAddress2": "",
          "billingAddressSameAsShippingAddress": true,
          "billingCity": "",
          "billingState": "",
          "billingZip": "",
          "cardholderName": "",
          "contactEmail": "",
          "contactName": "",
          "contactPhone": "",
          "shippingAddress1": "",
          "shippingAddress2": "",
          "shippingCity": "",
          "shippingCompanyName": "",
          "shippingName": "",
          "shippingState": "",
          "shippingZip": "",
          "storeBilling": false,
          "storeContactAndShipping": false,
        },
        "submissionError": null,
        "submitting": false,
        "updateStorageDisposer": null,
      }
    }
    submit={[Function]}
  />
</AppLayout>
`;
